crypto/fipsmodule/service_indicator/service_indicator_test.cc

```diff
@@ -63,6 +67,273 @@ static const uint8_t kAESCBCCiphertext[64] = {
+    { sha384, kPlaintext, kOutput_sha384,true },
+    { sha512, kPlaintext, kOutput_sha512,true },
+    { sha512_256, kPlaintext, kOutput_sha512_256,true }
+};
+
+class EVP_MD_ServiceIndicatorTest : public testing::TestWithParam<DigestTestVector> {};
+
+INSTANTIATE_TEST_SUITE_P(All, EVP_MD_ServiceIndicatorTest, testing::ValuesIn(kDigestTestVectors));
+
+TEST_P(EVP_MD_ServiceIndicatorTest, EVP_Ciphers) {
+  const DigestTestVector &t = GetParam();
+
+  int approved = AWSLC_NOT_APPROVED;
+  bssl::ScopedEVP_MD_CTX ctx;
+  std::unique_ptr<uint8_t[]> digest(new uint8_t[EVP_MD_size(t.md.func())]);
+  unsigned digest_len;
+
+  // Test the EVP_Digest interfaces for approval one by one directly.
+  CALL_SERVICE_AND_CHECK_APPROVED(approved, ASSERT_TRUE(EVP_DigestInit_ex(ctx.get(), t.md.func(), nullptr)));
+  if (t.expect_approved) {
+    ASSERT_TRUE(approved);
+  } else {
+    ASSERT_FALSE(approved);
+  }
+  CALL_SERVICE_AND_CHECK_APPROVED(approved, ASSERT_TRUE(EVP_DigestUpdate(ctx.get(), t.input, sizeof(t.input))));
+  if (t.expect_approved) {
+    ASSERT_TRUE(approved);
+  } else {
+    ASSERT_FALSE(approved);
+  }
+  CALL_SERVICE_AND_CHECK_APPROVED(approved, ASSERT_TRUE(EVP_DigestFinal_ex(ctx.get(), digest.get(), &digest_len)));
+  if (t.expect_approved) {
+    ASSERT_TRUE(check_test(t.expected_digest, digest.get(), digest_len, t.md.name));
+    ASSERT_TRUE(approved);
+  } else {
+    ASSERT_FALSE(approved);
+  }
+
+  // Test using the one-shot API for approval.
+  CALL_SERVICE_AND_CHECK_APPROVED(approved, t.md.one_shot_func(t.input, sizeof(t.input), digest.get()));
+  if (t.expect_approved) {
+    ASSERT_TRUE(check_test(t.expected_digest, digest.get(), EVP_MD_size(t.md.func()), t.md.name));
+    ASSERT_TRUE(approved);
+  } else {
+    ASSERT_FALSE(approved);
+  }
+}
+
+struct HMACTestVector {
+  // func is the hash function for HMAC to test.
+  const EVP_MD *(*func)(void);
+  // input is a NUL-terminated string to hash.
+  const uint8_t *input;
+  // expected_hex is the expected digest.
+  const uint8_t *expected_digest;
+  // expected to be approved or not.
+  const bool expect_approved;
+} kHMACTestVectors[] = {
+    { EVP_sha1, kPlaintext, kHMACOutput_sha1, true },
+    { EVP_sha224, kPlaintext, kHMACOutput_sha224, true },
+    { EVP_sha256, kPlaintext, kHMACOutput_sha256,true },
+    { EVP_sha384, kPlaintext, kHMACOutput_sha384,true },
+    { EVP_sha512, kPlaintext, kHMACOutput_sha512,true },
+    { EVP_sha512_256, kPlaintext, nullptr, false }
+};
+
+class HMAC_ServiceIndicatorTest : public testing::TestWithParam<HMACTestVector> {};
+
+INSTANTIATE_TEST_SUITE_P(All, HMAC_ServiceIndicatorTest, testing::ValuesIn(kHMACTestVectors));
+
+TEST_P(HMAC_ServiceIndicatorTest, HMACTest) {
+  const HMACTestVector &t = GetParam();
+
+  int approved = AWSLC_NOT_APPROVED;
+  const uint8_t kHMACKey[64] = {0};
+  std::string digest_str;
+  const EVP_MD *digest = t.func();
+  std::vector<uint8_t> key(kHMACKey, kHMACKey + sizeof(kHMACKey));
```

Comment:
I'm trying to decipher what this is doing? Is there some constructor somewhere I don't know off?